---
title: "Non-disclosive federated omic data analysis with DataSHIELD and Bioconductor"
author:
- name: Yannick Marcon
  affiliation:
  - &epigeny Epigeny
  email: yannick.marcon@obiba.org
- name: Juan R. Gonzalez
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
  - &uab Department of Mathematics, Autonomous University of Barcelona (UAB)
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
package: "`r pkg_ver('dsOmics')`"
output:
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    toc_float: true
bibliography: dsOmics_Vignette.bib
vignette: >
  %\VignetteIndexEntry{Epigenomic data analysis with federated data using DataSHIELD infrastructure}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(comment="", warning=FALSE, message=FALSE, cache=FALSE)
```

# Purpose

```{r child='rmd/purpose.Rmd', include=TRUE}
```

# Setup

```{r child='rmd/setup.Rmd', include=TRUE}
```

## Required Opal server with resources

```{r child='rmd/required_opal_servers.Rmd', include=TRUE}
```

## Required DataSHIELD packages in the opal server

```{r child='rmd/required_datashield_packages.Rmd', include=TRUE}
```

## Required R Packages in the client site (e.g. local machine)

```{r child='rmd/required_r_packages.Rmd', include=TRUE}
```

# Transcriptomic and Epigenomic data analysis

## Single study analysis

```{r child='rmd/single_study_analysis.Rmd', include=TRUE}
```

### Illustrative example: differential gene expression (DGE) analysis

```{r child='rmd/single_study_analysis_example.Rmd', include=TRUE}
```

## Analysis from multiple studies

```{r child='rmd/multiple_study_analysis.Rmd', include=TRUE}
```

### Illustrative example: Epigenome-wide association analysis (EWAS)

```{r child='rmd/multiple_study_analysis_example.Rmd', include=TRUE}
```

## Single CpG analysis

```{r child='rmd/single_cpg_analysis.Rmd', include=TRUE}
```

## Multiple CpG analysis

```{r child='rmd/multiple_cpg_analysis.Rmd', include=TRUE}
```

## Adjusting for Surrogate Variables

```{r child='rmd/surrogate_variables.Rmd', include=TRUE}
```

# Genomic data analysis: GWAS

```{r child='rmd/gwas.Rmd', include=TRUE}
```

## Analysis with Bioconductor packages: example how to extend the resources

```{r child='rmd/example_resources.Rmd', include=TRUE}
```

## GWAS with PLINK

```{r child='rmd/plink.Rmd', include=TRUE}
```

# Acknowledgments
JRG want to thank Deroshan Padotan for having worked on a preliminary version of `dsOmicsClient` package developed before the `resourcer` package was created. We also thank Demetris Avraam for his feedback while programming DataSHIELD functions and writing the vignette. Leire Abarrategui is acknowledged for implementing RNA-seq data analysis using DESeq2 and edgeR. Xavier Escrib√† is also acknowledged for the implementation of several functions to summarize and perform quality control of SNP data.

# References
